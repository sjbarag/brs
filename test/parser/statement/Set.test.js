const BrsError = require("../../../lib/Error");

const { Lexeme, BrsTypes, Parser } = require("brs");
const { Int32, BrsString } = BrsTypes;

const { EOF } = require("../ParserTests");
describe("parser indexed assignment", () => {
    it("assigns to dotted index", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.Identifier, text: "foo", line: 1 },
            { kind: Lexeme.Dot, text: ".", line: 1 },
            { kind: Lexeme.Identifier, text: "bar", line: 1 },
            { kind: Lexeme.Equal, text: "=", line: 1 },
            { kind: Lexeme.String, text: "baz", line: 1, literal: new BrsString("baz") },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });

    it("assigns to bracketed index", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.Identifier, text: "someArray", line: 1 },
            { kind: Lexeme.LeftSquare, text: "[", line: 1 },
            { kind: Lexeme.Integer, text: "0", line: 1, literal: new Int32(0) },
            { kind: Lexeme.RightSquare, text: "]", line: 1 },
            { kind: Lexeme.Equal, text: "=", line: 1 },
            { kind: Lexeme.String, text: "baz", line: 1, literal: new BrsString("baz") },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });
});
