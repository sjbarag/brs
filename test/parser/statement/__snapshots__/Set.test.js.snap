// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser indexed assignment bracketed assignment operator 1`] = `
Array [
  IndexedSet {
    "closingSquare": Object {
      "isReserved": false,
      "kind": "RightSquare",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "]",
    },
    "index": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 0,
      },
    },
    "obj": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "someArray",
      },
      "type": "Variable",
    },
    "type": "IndexedSet",
    "value": Binary {
      "left": IndexedGet {
        "closingSquare": Object {
          "isReserved": false,
          "kind": "RightSquare",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "]",
        },
        "index": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Int32 {
            "kind": 4,
            "value": 0,
          },
        },
        "obj": Variable {
          "name": Object {
            "isReserved": false,
            "kind": "Identifier",
            "literal": undefined,
            "location": Object {
              "end": Object {
                "column": -9,
                "line": -9,
              },
              "start": Object {
                "column": -9,
                "line": -9,
              },
            },
            "text": "someArray",
          },
          "type": "Variable",
        },
        "type": "IndexedGet",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 3,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "StarEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "*=",
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser indexed assignment bracketed assigns anonymous functions 1`] = `
Array [
  IndexedSet {
    "closingSquare": Object {
      "isReserved": false,
      "kind": "RightSquare",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "]",
    },
    "index": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 0,
      },
    },
    "obj": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "someArray",
      },
      "type": "Variable",
    },
    "type": "IndexedSet",
    "value": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "file": undefined,
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "statements": Array [],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndFunction",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end function",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Function",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "function",
      },
      "parameters": Array [],
      "returns": 10,
      "type": "Function",
    },
  },
]
`;

exports[`parser indexed assignment bracketed assigns boolean expressions 1`] = `
Array [
  IndexedSet {
    "closingSquare": Object {
      "isReserved": false,
      "kind": "RightSquare",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "]",
    },
    "index": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 0,
      },
    },
    "obj": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "someArray",
      },
      "type": "Variable",
    },
    "type": "IndexedSet",
    "value": Binary {
      "left": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": BrsBoolean {
          "kind": 2,
          "value": true,
        },
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": BrsBoolean {
          "kind": 2,
          "value": false,
        },
      },
      "token": Object {
        "isReserved": true,
        "kind": "And",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "and",
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser indexed assignment dotted assignment operator 1`] = `
Array [
  DottedSet {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "bar",
    },
    "obj": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "foo",
      },
      "type": "Variable",
    },
    "type": "DottedSet",
    "value": Binary {
      "left": DottedGet {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "bar",
        },
        "obj": Variable {
          "name": Object {
            "isReserved": false,
            "kind": "Identifier",
            "literal": undefined,
            "location": Object {
              "end": Object {
                "column": -9,
                "line": -9,
              },
              "start": Object {
                "column": -9,
                "line": -9,
              },
            },
            "text": "foo",
          },
          "type": "Variable",
        },
        "type": "DottedGet",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 5,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "StarEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "*=",
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser indexed assignment dotted assigns anonymous functions 1`] = `
Array [
  DottedSet {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "bar",
    },
    "obj": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "foo",
      },
      "type": "Variable",
    },
    "type": "DottedSet",
    "value": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "file": undefined,
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "statements": Array [],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndFunction",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end function",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Function",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "function",
      },
      "parameters": Array [],
      "returns": 10,
      "type": "Function",
    },
  },
]
`;

exports[`parser indexed assignment dotted assigns boolean expressions 1`] = `
Array [
  DottedSet {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "bar",
    },
    "obj": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "foo",
      },
      "type": "Variable",
    },
    "type": "DottedSet",
    "value": Binary {
      "left": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": BrsBoolean {
          "kind": 2,
          "value": true,
        },
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": BrsBoolean {
          "kind": 2,
          "value": false,
        },
      },
      "token": Object {
        "isReserved": true,
        "kind": "And",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "and",
      },
      "type": "Binary",
    },
  },
]
`;
